File: 数据结构js/maxHeap.js

// 使用数组存储二叉堆
//
// 二叉堆定义:
// * 完全二叉树
// * 子树小于父树
// * 左子树小于右子树

class MaxHeap {
	
	constructor( capacity ) {
	    this.data = []
	    this.count = 0
	}

	size() {
		return this.count
	}

	isEmpty() {
	  return this.count === 0
	}

	insert(item) {
		// data 数组下标从 1 开始
		this.data[++this.count] = item

		// shiftUp 用来维护堆的定义
		__shiftUp.call(this, this.count )

	   function __shiftUp(index) {
			while( index > 1 && this.data[index] > this.data[~~(index/2)]) {
				let t = this.data[index]
				this.data[index] = this.data[~~(index/2)]
				this.data[~~(index/2)] = t
				index = ~~(index/2)
		    }
	    }
	}

	extractMax() {
		let res = this.data[1]
		this.data[1] = this.data[this.count]
		this.data[this.count] = res
		this.count--
		this.data.length--

		__shiftDown.call(this, 1)
		return res

		function __shiftDown(index) {
			while( index * 2 < this.count ) {
				let j = index * 2
				if( j + 1 < this.count &&  this.data[j + 1] > this.data[j] )
					j += 1
				if( this.data[index] >= this.data[j] )
					break

				let t = this.data[index]
				this.data[index] = this.data[j]
				this.data[j] = t
				index = j
			}
		}
	}
}

let maxheap = new MaxHeap(100)

for(var i = 0; i < 10; i++) {
	maxheap.insert(~~(Math.random() * 100) )
}
console.log(maxheap.data)
console.log(maxheap.extractMax())

console.log(maxheap.data)

File: java/src/main/java/com/java/algorithm/heap/MaxHeap.java
package com.java.algorithm.heap;

public class MaxHeap<T extends Comparable> {
  private T data[];
  private int size;
  private int capacity;

  private void __shiftUp(int k) {
    int parentIndex = k/2;
    while (k > 1 && data[k].compareTo(data[parentIndex]) > 0) {
      __swap(data, k, parentIndex);
      k = k/2;
    }
  }

  private void __swap(T[] arr, int i, int j) {
    T temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
  }

  private void __shiftDown(int k) {
    while (2*k <= size) {
      int j = 2*k; // 在本次循环中， data[k] 和 data[j] 交换位置

      // 有右孩子，且有孩子大于左孩子
      if (j + 1 <= size && data[j+1].compareTo(data[j]) > 0) j += 1;

      if (data[k].compareTo(data[j]) >= 0) break;

      __swap(data, k, j);
      k = j;
    }
  }

  MaxHeap(int capacity) {
    this.capacity = capacity;
    size = 0;
    data = (T[]) new Object[capacity];
  }

  public void insert(T item) {
    assert (size + 1 <= capacity);
    data[size + 1] = item;
    size++;

    __shiftUp(size);
  }

  public T extractMax() {
    assert (size > 0);

    T max = data[1];
    data[1] = data[size];
    --size;

    __shiftDown(1);

    return max;
  }

  public int size() {
    return size;
  }

  public boolean isEmpty() {
    return size == 0;
  }
}
